### ISSUE TITLE
`GlooBackend` leaks file descriptors

### ISSUE THREAD

**Issue by NeoLegends** [pytorch/pytorch#129868]
### 🐛 Describe the bug

In RETURNN we ran into issues on long running trainings wrt. `too many open files`. Upon closer inspection we found the `DataLoader` initializes a new new process group for every iteration through the dataset [^1] to synchronize the random seed between workers. It seems like this process group uses some pipes and sockets to coordinate, many of which are never torn down again, even when the iterator goes out of scope.

A workaround is using `num_workers > 0` and setting `persistent_workers=True`. Then the iteratior and its process group is only created once and the object is reused for subsequent iterations.

Related issue: https://github.com/rwth-i6/returnn/issues/1560

Code to reproduce:

```python
import gc
import os
import subprocess
import time

import torch.distributed as dist


if __name__ == "__main__":
    assert dist.is_available()

    dist.init_process_group()

    for i in range(10):
        grp = dist.new_group(backend="gloo")

        cmd = ["lsof", "-l", "-p", str(os.getpid())]
        out = subprocess.run(cmd, capture_output=True, text=True)

        print(
            f"Process {dist.get_rank()}, iteration {i + 1}:",
            len(out.stdout.split("\n")),
            "open FDs",
        )

        del grp
        gc.collect()

        # FDs are not released, even though group should have been torn down?

        time.sleep(0.1)
```

save as `test.py`, run via

```bash
#!/usr/bin/env bash
python3 -mtorch.distributed.run --standalone --nnodes=1 --nproc-per-node=2 ./test.py
```

Output (sorted by process):

```
Process 0, iteration 1: 97 open FDs
Process 0, iteration 2: 102 open FDs
Process 0, iteration 3: 107 open FDs
Process 0, iteration 4: 112 open FDs
Process 0, iteration 5: 117 open FDs
Process 0, iteration 6: 122 open FDs
Process 0, iteration 7: 127 open FDs
Process 0, iteration 8: 132 open FDs
Process 0, iteration 9: 137 open FDs
Process 0, iteration 10: 142 open FDs
Process 1, iteration 1: 92 open FDs
Process 1, iteration 2: 97 open FDs
Process 1, iteration 3: 102 open FDs
Process 1, iteration 4: 107 open FDs
Process 1, iteration 5: 112 open FDs
Process 1, iteration 6: 117 open FDs
Process 1, iteration 7: 122 open FDs
Process 1, iteration 8: 127 open FDs
Process 1, iteration 9: 132 open FDs
Process 1, iteration 10: 137 open FDs
```

[^1]: https://github.com/pytorch/pytorch/blob/f6a0be5023d5bbd5b6e6248d13326b13d42c9f8d/torch/utils/data/dataloader.py#L629-L630

### Versions

Collecting environment information...
PyTorch version: 2.3.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-107-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             96
On-line CPU(s) list:                0-95
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8168 CPU @ 2.70GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          2
Stepping:                           4
BogoMIPS:                           5400.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           48 MiB (48 instances)
L3 cache:                           66 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94
NUMA node1 CPU(s):                  1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS; IBPB conditional; STIBP conditional; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT vulnerable

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.16.1
[pip3] onnxruntime==1.18.0
[pip3] torch==2.3.1
[pip3] triton==2.3.1
[conda] Could not collect

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k

**Comment by XilunWu** [comment]
`persistent_workers=True` is the recommended setting. Backward compatibility is the reason why we keep it by default `False`.
